src/main/cpp/src/timezones.cu

```diff
@@ -0,0 +1,206 @@
+                                                     lists_column_device_view const &transitions,
+                                                     size_type tz_index, bool to_utc) {
+
+  using duration_type = typename timestamp_type::duration;
+  using cuda::std::chrono::duration_cast;
+
+  auto epoch_seconds =
+      static_cast<int64_t>(duration_cast<cudf::duration_s>(timestamp.time_since_epoch()).count());
+
+  auto const tz_transitions = cudf::list_device_view{transitions, tz_index};
+
+  auto size = tz_transitions.size();
+
+  cudf::device_span<int64_t const> transition_times(
+      &(transitions.child()
+            .child(to_utc ? 1 : 0)
+            .data<int64_t>()[tz_transitions.element_offset(0)]),
+      static_cast<size_t>(size));
+
+  auto idx = thrust::upper_bound(thrust::seq, transition_times.begin(), transition_times.end(),
+                                 epoch_seconds) -
+             transition_times.begin();
+
+  auto const list_offset = tz_transitions.element_offset(size_type(idx - 1));
+  auto const utc_offset = duration_cast<duration_type>(cudf::duration_s{
+      static_cast<int64_t>(transitions.child().child(2).element<int32_t>(list_offset))});
+  return to_utc ? timestamp - utc_offset : timestamp + utc_offset;
+}
+
+} // namespace
+
+namespace spark_rapids_jni {
+
+std::unique_ptr<column> convert_timestamp_to_utc(cudf::column_view const &input,
+                                                 cudf::table_view const &transitions,
+                                                 size_type tz_index, rmm::cuda_stream_view stream,
+                                                 rmm::mr::device_memory_resource *mr) {
+
+  auto type = input.type().id();
+  auto num_rows = input.size();
+
+  // get the fixed transitions
+  auto const ft_cdv_ptr = column_device_view::create(transitions.column(0), stream);
+  lists_column_device_view fixed_transitions = cudf::detail::lists_column_device_view{*ft_cdv_ptr};
+
+  auto results = cudf::make_timestamp_column(input.type(), input.size(),
+                                             cudf::detail::copy_bitmask(input, stream, mr),
+                                             input.null_count(), stream, mr);
+
+  switch (type) {
+    case cudf::type_id::TIMESTAMP_SECONDS:
+      thrust::transform(rmm::exec_policy(stream), thrust::make_counting_iterator(0),
+                        thrust::make_counting_iterator(num_rows),
+                        results->mutable_view().begin<cudf::timestamp_s>(),
+                        [input_data = input.begin<cudf::timestamp_s>(), fixed_transitions,
+                         tz_index] __device__(auto const i) {
+                          auto const timestamp = input_data[i];
+                          return convert_timestamp_timezone<cudf::timestamp_s>(
+                              timestamp, fixed_transitions, tz_index, true);
+                        });
```

Comment:
Recommend to follow what is suggested here: https://github.com/NVIDIA/spark-rapids-jni/pull/1553#discussion_r1396481946